path.go

```diff
@@ -370,3 +479,134 @@ func RemoveEscapeChar(word string) string {
 	}
 	return word
 }
+
+func getParamConstraintType(constraintPart string) TypeConstraint {
+	switch constraintPart {
+	case "int":
+		return intConstraint
+	case "bool":
+		return boolConstraint
+	case "float":
+		return floatConstraint
+	case "alpha":
+		return alphaConstraint
+	case "guid":
+		return guidConstraint
+	case "minLen":
+		return minLengthConstraint
+	case "maxLen":
+		return maxLengthConstraint
+	case "exactLen":
+		return exactLengthConstraint
+	case "betweenLen":
+		return BetweenLengthConstraint
+	case "min":
+		return minConstraint
+	case "max":
+		return maxConstraint
+	case "range":
+		return rangeConstraint
+	case "datetime":
+		return datetimeConstraint
+	case "regex":
+		return regexConstraint
+	default:
+		return noConstraint
+	}
+
+}
+
+func (c *Constraint) CheckConstraint(param string, matcher utils.RegexMatch) bool {
+	var err error
+	var num int
+
+	// check data exists
+	needOneData := []TypeConstraint{minLengthConstraint, maxLengthConstraint, exactLengthConstraint, minConstraint, maxConstraint, datetimeConstraint, regexConstraint}
+	needTwoData := []TypeConstraint{BetweenLengthConstraint, rangeConstraint}
+
+	for _, data := range needOneData {
+		if c.ID == data && len(c.Data) == 0 {
+			return false
+		}
+	}
+
+	for _, data := range needTwoData {
+		if c.ID == data && len(c.Data) < 2 {
+			return false
+		}
+	}
+
+	switch c.ID {
+	case intConstraint:
+		_, err = strconv.Atoi(param)
+	case boolConstraint:
+		_, err = strconv.ParseBool(param)
+	case floatConstraint:
+		_, err = strconv.ParseFloat(param, 32)
+	case alphaConstraint:
+		for _, r := range param {
+			if !unicode.IsLetter(r) {
+				err = errors.New("")
+			}
+		}
+	case guidConstraint:
+		_, err = uuid.Parse(param)
+	case minLengthConstraint:
+		data, _ := strconv.Atoi(c.Data[0])
+
+		if len(param) < data {
+			err = errors.New("")
+		}
+	case maxLengthConstraint:
+		data, _ := strconv.Atoi(c.Data[0])
+
+		if len(param) > data {
+			err = errors.New("")
+		}
+	case exactLengthConstraint:
+		data, _ := strconv.Atoi(c.Data[0])
+
+		if len(param) != data {
+			err = errors.New("")
+		}
+	case BetweenLengthConstraint:
+		data, _ := strconv.Atoi(c.Data[0])
+		data2, _ := strconv.Atoi(c.Data[1])
+		length := len(param)
+
+		if !(length >= data && length <= data2) {
+			err = errors.New("")
+		}
+	case minConstraint:
+		data, _ := strconv.Atoi(c.Data[0])
+		num, err = strconv.Atoi(param)
+
+		if num < data {
+			err = errors.New("")
+		}
+	case maxConstraint:
+		data, _ := strconv.Atoi(c.Data[0])
+		num, err = strconv.Atoi(param)
+
+		if num > data {
+			err = errors.New("")
+		}
+	case rangeConstraint:
+		data, _ := strconv.Atoi(c.Data[0])
+		data2, _ := strconv.Atoi(c.Data[1])
+		num, err = strconv.Atoi(param)
+
+		if !(num >= data && num <= data2) {
+			err = errors.New("")
+		}
+	case datetimeConstraint:
+		_, err = time.Parse(c.Data[0], param)
+	case regexConstraint:
+		match, _ := matcher(c.Data[0], param)
```

Comment:
precompiled matchers are faster

when writing the contrains into the route, we should take the regex based on the passed regex matcher in the app configuration and precompile the regular expression, so that the additional parameter which is injected via the "match" can be omitted and by precompiling (which happens when building the routes -> so is not performance critical) the performance increases

Reply:
yes. it's good idea to use Compile for performance. But it'll be difficult to add customization for regex section it after using Compile. It seems there's a deal :smiley: